source/Subsystem/JobSystem.cpp

```diff
@@ -1,103 +1,120 @@
+//	return true;
+//};
 
-void JobSystem::Terminate()
+bool JobSystem::Init()
 {
-	CollectAllThreads();
+	// Example for adding jobs:
+	//for (int i = 0; i < 30; i++)
+	//	AddJob(exampleJob);
+
+	return true;
 }
 
-bool JobSystem::CollectAllThreads()
+void JobSystem::Update(float aFrameTime)
 {
-	for (Job& job : myJobs)
+	// First lets check if any futures have finished.
+	for (int i = myRunningJobs.size() - 1; i >= 0; --i)
 	{
-		// If future is valid, so is the task.
-		if (job.myFuture.valid())
+		Job& job = myRunningJobs[i];
+		const std::future_status status = job.myFuture.wait_for(std::chrono::seconds(0));
+		if (status == std::future_status::ready)
 		{
-			job.myFuture.wait();
-			if (job.myThread.joinable())
+			Log::Print(LogType::MESSAGE, "A task finished with result %i!", job.myFuture.get());
+
+			for (int i = 0; i < myThreads.size(); ++i)
 			{
-				job.myThread.join();
-				job = Job();
+				std::thread* threadToRemove = myThreads[i];
+				if (threadToRemove == *job.myThread)
+				{
+					threadToRemove->join();
+					delete(threadToRemove);
+					myThreads[i] = nullptr;
+					break;
+				}
 			}
+
+			job.myThread = nullptr;
+			myRunningJobs.erase(myRunningJobs.begin() + i);
+			myNeedsUpdate = true;
 		}
 	}
-	
-	return true;
-}
-
-int JobSystem::CollectOneThread()
-{
-	// Wait for a thread to open up.
-	const size_t jobCount = myJobs.size();
-	const bool hasJobs = jobCount != 0;
-	std::chrono::nanoseconds waitForThreadTime(1);
 
-	Timer expireTimer;
-	expireTimer.Start();
-
-	while (hasJobs)
+	// Add jobs to the queue.
+	if (myNeedsUpdate && !myQueuedJobs.empty())
 	{
-		for (size_t i = 0; i < jobCount; ++i)
+		// Any threads not in use?
+		std::vector<std::thread**> emptyThreads;
+
+		int emptyThreadCount = 0;
+		for (int threadIndex = 0; threadIndex <= myThreads.capacity() - 1 && emptyThreadCount <= myQueuedJobs.size() - 1; ++threadIndex)
 		{
-			if (!myJobs[i].myFuture.valid())
+			if (myThreads[threadIndex] == nullptr)
 			{
-				return static_cast<int>(i);
-			}
-			if (myJobs[i].myFuture.wait_for(waitForThreadTime) == std::future_status::ready)
-			{
-				myJobs[i].myThread.join();
-				return static_cast<int>(i);
+				emptyThreads.push_back(&myThreads[threadIndex]);
+				emptyThreadCount++;
 			}
 		}
-		if (expireTimer.GetCurrentTime() > myMaximumExpirationTime)
+
+		// Put all jobs we can start in there.
+		for (int i = emptyThreads.size() - 1; i >= 0 && emptyThreadCount > 0; --i)
 		{
-			expireTimer.Reset();
-			Log::Print(LogType::WARNING, "Threads all being locked for a long time. If unintended, consider closing the engine");
+			Job& job = myQueuedJobs[myQueuedJobs.size() - i - 1];
+			StartJob(job, *emptyThreads[i]);
+			myQueuedJobs.erase(myQueuedJobs.end() - i - 1);
+			emptyThreadCount--;
 		}
 	}
 
-	Assert(true, "Job count is zero! This should never happen!");
+	myNeedsUpdate = false;
+	return;
+}
+
+bool JobSystem::Terminate()
+{
+	while (!myRunningJobs.empty())
+		Update(0.016f);
```

Comment:
If this is required for the JobSystem then it might be worth adding a good solution for this in the architecture branch when terminating systems (after this commit).